src/evaluation/context.rs

```diff
@@ -0,0 +1,46 @@
+use std::collections::HashMap;
+
+use typed_builder::TypedBuilder;
+
+use crate::EvaluationContextFieldValue;
+
+/// The evaluation context provides ambient information for the purposes of flag evaluation.
+/// Contextual data may be used as the basis for targeting, including rule-based evaluation,
+/// overrides for specific subjects, or fractional flag evaluation.
+///
+/// The context might contain information about the end-user, the application, the host, or any
+/// other ambient data that might be useful in flag evaluation. For example, a flag system might
+/// define rules that return a specific value based on the user's email address, locale, or the
+/// time of day. The context provides this information. The context can be optionally provided at
+/// evaluation, and mutated in before hooks.
+#[derive(Clone, TypedBuilder, Default)]
+pub struct EvaluationContext {
+    /// The targeting key uniquely identifies the subject (end-user, or client service) of a flag
+    /// evaluation. Providers may require this field for fractional flag evaluation, rules, or
+    /// overrides targeting specific users. Such providers may behave unpredictably if a targeting
+    /// key is not specified at flag resolution.
+    #[builder(default, setter(into, strip_option))]
+    pub targeting_key: Option<String>,
+
+    #[builder(default)]
+    pub custom_fields: HashMap<String, EvaluationContextFieldValue>,
+}
+
+impl EvaluationContext {
+    pub fn with_custom_field<S: Into<String>, V: Into<EvaluationContextFieldValue>>(
+        mut self,
+        key: S,
+        value: V,
+    ) -> Self {
+        self.add_custom_field(key, value);
+        self
+    }
+
+    pub fn add_custom_field<S: Into<String>, V: Into<EvaluationContextFieldValue>>(
```

Comment:
Is there a reason the user would call this instead of the one above? I think if they don't care about the return value, then they just don't use it, which would simplify this API.

Or maybe the one above is intended to be non-mutating?

Reply:
The `with_custom_field` takes a `mut Self` and returns a `Self`. It is designed to be chained. But if the client code has already constructed an instance, calling `with_custom_field` on that instance will take the ownership. For example:

```rust
let context = EvaluationContext::default();
context.with_custom_field(...);

// You cannot use context anymore because its ownership has been taken by the `with_xx` function.
// In order to do so, you have to say:
// context = context.with_custom_field(...);
```

I think it is not so convenient considering the custom field might be added lately, so both ways are provided.

You can check [derive_builder's build patterns](https://docs.rs/derive_builder/latest/derive_builder/#builder-patterns) for more details about tradeoffs among different ways.